components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java

```diff
@@ -553,6 +602,29 @@ Some claims (Eg:- Birth date) can have "/" in claim values. But in user store le
         }
     }
 
+    /**
+     * Get all the domain names related to user stores.
+     * @param tenantId   Tenant ID
+     * @return A list of all the available domain names for given tenant
+     */
+    private List<String> getDomainNames(int tenantId) throws IdentityRecoveryServerException {
+
+        List<String> domainsOfUserStores = new ArrayList<>();
+        AbstractUserStoreManager abstractUserStoreManager = (AbstractUserStoreManager) getUserStoreManager(tenantId);
+        UserStoreManager secondaryUserStore = abstractUserStoreManager.getSecondaryUserStoreManager();
+        while (secondaryUserStore != null) {
+            String domainName = secondaryUserStore.getRealmConfiguration().
+                    getUserStoreProperty(UserCoreConstants.RealmConfig.PROPERTY_DOMAIN_NAME).toUpperCase();
+            secondaryUserStore = secondaryUserStore.getSecondaryUserStoreManager();
+            domainsOfUserStores.add(domainName);
+        }
+
+        /* Append the primary domain name to the front of the domain list since the first iteration of multiple
+           domain filtering should happen for the primary user store. */
+        domainsOfUserStores.add(0, UserCoreConstants.PRIMARY_DEFAULT_DOMAIN_NAME);
```

Comment:
We can do this in the beginning, that would avoid the need for shifting other elements in the list

Reply:
Resolved by https://github.com/wso2-extensions/identity-governance/pull/623/commits/1e25e83cae63fad1bc4721331a7e4ab887046cd6